numpy/core/src/npysort/radixsort.c.src

```diff
@@ -0,0 +1,236 @@
+#define NPY_NO_DEPRECATED_API NPY_API_VERSION
+
+#include "npy_sort.h"
+#include "npysort_common.h"
+#include <numpy/npy_endian.h>
+#include <stdlib.h>
+
+/*
+ *****************************************************************************
+ **                            INTEGER SORTS                                **
+ *****************************************************************************
+ */
+
+
+/**begin repeat
+ *
+ * #TYPE = BOOL, BYTE, UBYTE, SHORT, USHORT, INT, UINT, LONG, ULONG,
+ *         LONGLONG, ULONGLONG#
+ * #suff = bool, byte, ubyte, short, ushort, int, uint, long, ulong,
+ *         longlong, ulonglong#
+ * #type = npy_ubyte, npy_ubyte, npy_ubyte, npy_ushort, npy_ushort, npy_uint,
+ *         npy_uint, npy_ulong, npy_ulong, npy_ulonglong, npy_ulonglong#
+ * #sign = 0, 1, 0, 1, 0, 1, 0, 1, 0, 1, 0#
+ * #floating = 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0#
+ */
+
+#if NPY_BYTE_ORDER == NPY_LITTLE_ENDIAN
+    #if @sign@
+        #if @floating@
+            #define KEY_OF(x) ((x) ^ (-((x) >> (sizeof(@type@) * 8 - 1)) | ((@type@)1 << (sizeof(@type@) * 8 - 1))))
+        #else
+            #define KEY_OF(x) ((x) ^ ((@type@)1 << (sizeof(@type@) * 8 - 1)))
+        #endif
+    #else
+        #define KEY_OF(x) (x)
+    #endif
+    #define PULL_BYTE(k, l) (((k) >> ((l) << 3)) & 0xFF)
+#else
+    #if @sign@
+        #if @floating@
+            #define KEY_OF(x) ((x) ^ (-((x) << (sizeof(@type@) * 8 - 1)) | ((@type@)1 >> (sizeof(@type@) * 8 - 1))))
+        #else
+            #define KEY_OF(x) ((x) ^ ((@type@)1 >> (sizeof(@type@) * 8 - 1)))
+        #endif
+    #else
+        #define KEY_OF(x) (x)
+    #endif
+    #define PULL_BYTE(k, l) (((k) << ((l) >> 3)) & 0xFF)
```

Comment:
I don't understand why this depends on endianness. This one looks wrong to me, as it return the lsbyte for all k < 8

Reply:
I reversed the bit-shifts in all cases. I may be wrong here but isn't that what's supposed to be done usually in cases of reversed endian-ness? `k` is the overall key with all bytes and I'm extracting a byte from it.